[issue-2] Resurrect Sprightly

OsidAbu-alrub opened the pull request.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,3 +1,60 @@
+): Promise<string> {
+  const regex = /<<(.*?)>>|\{\{(.*?)\}\}/gi
+  const file = (await readFile(entryPoint)).toString()
+  const result = file.replaceAll(regex, (match) => {
```
So ... the way you handled parsing the file is pretty neat. Though here's a suggestion for an implementation that might seem better:

As the callback for the `.replaceAll()` function, you should have this (this callback to `.replaceAll()` assumes that the regex is the one suggested above, you might have to modify it if you decided to go with a different one):
```js
(_, isComponent: string, reference: string) => {
  if (isComponent) {
    // logic to parse component
    return
  }

  /** input-output samples that explain how the function works
  * "key" => ["key"]
  * "property.nested[0].anotherProperty[1]" => ["property", "nested", 0, "anotherProperty", 1]
  */
  const path = subscriptingToPath(reference)

  // rest of logic to read the property using the path. I would suggest using `Array.reduce`, it's perfect for this situation.
}
```

A note about `subscriptingToPath` implementation:
Investigate [this regex](https://regex101.com/r/f4Y8zW/1) that could reduce the whole implementation of this function into 1 line, toggle between the different flavors of C# and JS, observe the Match Information widget, and try not to hate JS 😃.

obadakhalili replied:
https://regex101.com/r/KjfuU4/1

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,3 +1,60 @@
+export = async function sprightly(
+  entryPoint: string,
+  data: Data,
+  options: Options = defaultOptions,
```
This handles not passing the options parameter but doesn't handle passing an options object but not all options. You should merge the passed options parameter with the default options object.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,3 +1,60 @@
+      return ` ${match.substring(1, match.length - 1)}`
+    })
+    .split(" ")
+  return parsedKeys.reduce<any>((acc, key) => {
```
Nice one using reduce here.

obadakhalili reviewed (commented):
* You still need to implement the following:
  - `throwOnKeyNotfound`.
  - Caching.
* Please don't forget to amend the current commit when you have addressed these comments.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,60 +1,85 @@
+        .filter((v) => v && Boolean(v.trim()))
+        .reduce((obj, key) => {
+          if (
+            typeof obj === "object" &&
```
Why do you need to check if this is an object?

OsidAbu-alrub replied:
If it is a primitive property, and I use `in` against it, it will throw an error

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,60 +1,85 @@
+          if (
+            typeof obj === "object" &&
+            !(key in obj) &&
+            options.throwOnKeyNotfound
```
Move this before checking if the key exists to leverage short-circuiting.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,60 +1,85 @@
+          )
+            throw new KeyNotFoundError(key.trim())
+          return (obj[key] || options.keyFallback) as Data
+        }, data) as unknown as string
```
This is actually not necessary. The final reduced file could still be an object.

obadakhalili replied:
Maybe you need to handle this case where the "leaf" isn't a primitive value (string or number)?

OsidAbu-alrub replied:
Got any ideas? or do you have anything specific in mind to print out if the reduced value is not a primitive?

obadakhalili reviewed (changes requested):
* I think if you run `pnpm install` you will find some new changes made to `pnpm-lock.yaml`, these should be pushed already, shouldn't they?
* If you try to install the package from your test project, your will get an error indicating that husky isn't found, this is because your package manager is trying to run the `postinstall` script found at Sprightly's package.json, though this should only run when we install the package deps to work on it, not when the user tries to install it. Do you think you can find a solution for that?
* Please use new lines when appropriate 😄. Examples: After L34, before and after if statements at L62, etc.
* Please follow this https://twitter.com/dan_abramov/status/1570426739591114755?t=RkDd111ieOGZ-V22bQ1Ikg&s=19 😛.

OsidAbu-alrub commented on tsconfig.json:
```diff
@@ -1,13 +1,14 @@
     "types": ["node"],
-    "outDir": "./dist"
+    "outDir": "./dist",
+    "lib": ["ESNext"]
```
https://www.typescriptlang.org/tsconfig#lib

It doesn't accept target tho

obadakhalili commented on src/sprightly.ts:
```diff
@@ -58,28 +83,42 @@ function sprightly(
-  if (options.cache && cache.has(entryPointAbsolutePath))
-    return cache.get(entryPointAbsolutePath) as string
+
+  console.log(path.resolve(entryPoint))
```
Forgot a log here.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -58,28 +83,42 @@ function sprightly(
-    return cache.get(entryPointAbsolutePath) as string
+
+  console.log(path.resolve(entryPoint))
+  const entryPointAbsolutePath = path.resolve(path.normalize(entryPoint))
```
I think you forgot to address this https://github.com/obadakhalili/sprightly/pull/4#discussion_r978353593.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -58,28 +83,42 @@ function sprightly(
 }
-export = (entryPoint: string, data: Data, options: Options = defaultOptions) =>
-  new Promise<string>((resolve, reject) => {
+export = function sprightlyAsync(
```
Add a line above this, and add a separate `export = sprightlyAsync` please.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -58,28 +83,42 @@ function sprightly(
+    if (error instanceof PathNotFoundError) {
+      throw new PathNotFoundError(error.component, entryPoint)
+    }
     throw error
```
Here, instead of doing all the above, check if the error is thrown by `sprightlyAsync`, and if so, rethrow it using the yet-to-implement custom error class about a path being not found. If it is not, then just re-throw the error directly.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,54 +1,79 @@
+  readonly key: string
+  readonly src: string
+  constructor(key: string, src: string) {
+    const message = `Key "${key}" not found in "data" at ${src}`
```
No need for "in data". Only makes it more ambiguous.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,60 +1,85 @@
+    try {
+      resolve(sprightly(entryPoint, data, options))
+    } catch (error) {
+      reject(error as string)
```
I think you forgot this.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,54 +1,79 @@
-  cache: false,
-}
-
 function parse(file: string, data: Data, options: Options) {
```
You need to add the `filename` parameter to this that is mentioned here https://github.com/obadakhalili/sprightly/pull/4#discussion_r978353593.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,54 +1,79 @@
           return sprightly(reference, data, options)
         } catch (e) {
-          throw new PathNotFoundError(reference.trim())
+          throw new PathNotFoundError(reference, "")
```
Here you should the `filename parameter`.

obadakhalili reviewed (changes requested):
* For the future, please don't describe only one of the things you did in the commit message. It is fine now though because when the review is done I'm gonna ask you to squash all the commits into a single commit.
* If I run `pnpm install` it adds changes to `pnpm-lock.yaml`. Does it do the same for you? If so can you add them.
* https://github.com/obadakhalili/sprightly/pull/4#discussion_r979469187. Here I mean that the `lib` property should simply be moved to after `target` property.
* https://github.com/obadakhalili/sprightly/pull/4#discussion_r979465988. Maybe an error or the stringified object. Though it might be unserializable. I guess an error is an optimal choice. What do you think?

OsidAbu-alrub commented:
- I've addressed all the comments in the last two commits I pushed
- I've removed error classes and just used the generic Error class. And made it so that we don't need to identify the error anymore, we can just rethrow it. This was possible due to the fact that we passed the entry point to the `parse` function
- I exported both `sprightly` and `sprightlyAsync`
- I've handled all the cases where we have nested paths, it is working correctly now.
- There's nothing being added to my `pnpm.lock` file when I run `pnpm install`
- I'm facing no issues with Husky

OsidAbu-alrub commented:
Thank you for the great review معلم 🥰

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
-          throw new PathNotFoundError(reference, "")
+        // if last segement of path is a file
+        // we go back one step to get the directory of the file
+        if (path.extname(entryPoint)) {
```
I don't think this check is necessary. In fact, you can remove these whole 3 lines and replace them with the code I suggested in the next block.

OsidAbu-alrub replied:
I found out what the issue was, I was modifying `entryPoint`, which at the first iteration gave out a directory. But the other iterations got a directory at the last segment of the path (instead of a file name).

Consequences of side effects, crazy eh?

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
 }
 
-function parse(file: string, data: Data, options: Options) {
+function parse(file: string, entryPoint: string, data: Data, options: Options) {
```
I think `entryPoint` should be `filename`, and `file` should become `fileContent`.

OsidAbu-alrub replied:
I changed `file` to `fileContent`, but `entryPoint` is a path to a file, this is why I called it `entryPoint`. I changed it to filePath tho

obadakhalili replied:
Aha, nice.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
   const parsedFile = file.replaceAll(
     /\{\{(>?)(.*?)\}\}/g,
     (_, isComponent: string, reference: string) => {
       reference = reference.trim()
```
A new line here 😄

obadakhalili replied:
Hey e.e

obadakhalili replied:
Add a line break here e.e

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
+          entryPoint = path.normalize(path.join(entryPoint, ".."))
         }
+
+        const componentAbsolutePath = path.resolve(
```
You can remove the above and do this here: `path.join(path.dirname(filename), path.normalize(reference))`.

obadakhalili replied:
Here you can assume that parse only gets absolute file names. It is like its part of its API. unlike `sprightly`, which might get relative entry points or absolute ones and it has to handle both. Again, it is like its part of its API.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
       }
 
-      return reference
+      const parsedData = reference
```
I think `keyValue` might be a better name here. What do you think?

obadakhalili commented on src/sprightly.ts:
```diff
@@ -56,11 +47,17 @@ function parse(file: string, data: Data, options: Options) {
-        }, data) as unknown as string
+        }, data)
+
+      if (parsedData && typeof parsedData !== "string") {
```
Also, another allowed primitive value is "number".

obadakhalili commented on src/sprightly.ts:
```diff
@@ -1,52 +1,43 @@
+        )
+
+        if (!existsSync(componentAbsolutePath)) {
+          throw new Error(
```
I think you should create a base `SprightlyError` class so that users can distinguish errors thrown by us.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -118,7 +116,9 @@ export = function sprightlyAsync(
       resolve(sprightly(entryPoint, data, options))
     } catch (error) {
-      reject(error as string)
+      reject((error as Error).message)
```
Here reject `error as SprightlyError`.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -84,32 +81,33 @@ function sprightly(
     }
+
+    return parsedFile
+  } catch (error) {
```
No need to catch anything here anymore.

obadakhalili reviewed (changes requested):
Great work.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -11,33 +11,41 @@ interface Options {
-        }
+        const fileDirectory = path.normalize(path.join(filePath, ".."))
 
         const componentAbsolutePath = path.resolve(
```
Here you can `.join` because you can assume that the resulting path is an absolute one. You want to make this assumption because its part of the `parse` function API to deal with absolute paths, so if this assertion fails, the code will **fail fast**,

obadakhalili commented on src/sprightly.ts:
```diff
@@ -11,33 +11,41 @@ interface Options {
-        if (path.extname(entryPoint)) {
-          entryPoint = path.normalize(path.join(entryPoint, ".."))
-        }
+        const fileDirectory = path.normalize(path.join(filePath, ".."))
```
No need to normalize and you can assume that the path is valid. Also, you can use `path.dirname()` directly. Also, `fileDir` is shorter. Or it might be even enough to use it directly inline without defining a variable.

obadakhalili replied:
Also, when using the `dirname` util you will no longer need the comment 😄.

obadakhalili commented on src/sprightly.ts:
```diff
@@ -116,7 +124,7 @@ function sprightlyAsync(
       resolve(sprightly(entryPoint, data, options))
     } catch (error) {
-      reject((error as Error).message)
+      reject((error as SprightlyError).message)
```
Throw the error, not the message. This way the user can get the message and other info as well.
